packages/aws-cdk-lib/aws-ecr/lib/repository.ts

```diff
@@ -881,10 +877,8 @@ export class Repository extends RepositoryBase {
           'ecr:ListImages',
           'ecr:ListTagsForResource',
         ],
-        Resource: Lazy.list({ produce: () => repoArns }),
+        Resource: [`arn:${Aws.PARTITION}:ecr:${Stack.of(this).region}:${Stack.of(this).account}:repository/*`],
       });
-    } else {
-      (provider as any)[REPO_ARN_SYMBOL].push(this._resource.attrArn);
     }
```

Comment:
@rix0rrr 

Are the following changes correct?

before
```ts
    if (firstTime) {
      const repoArns = [this._resource.attrArn];
      (provider as any)[REPO_ARN_SYMBOL] = repoArns;

      // Use a iam policy to allow the custom resource to list & delete
      // images in the repository and the ability to get all repositories to find the arn needed on delete.
      // We lazily produce a list of repositories associated with this custom resource provider.
      provider.addToRolePolicy({
        Effect: 'Allow',
        Action: [
          'ecr:BatchDeleteImage',
          'ecr:DescribeRepositories',
          'ecr:ListImages',
          'ecr:ListTagsForResource',
        ],
        Resource: Lazy.list({ produce: () => repoArns }),
      });
    } else {
      (provider as any)[REPO_ARN_SYMBOL].push(this._resource.attrArn);
    }
```

after
```ts
    const role = iam.Role.fromRoleArn(this, `${this.repositoryName}Role`, provider.roleArn);

    // Use a iam policy to allow the custom resource to list & delete
    // images in the repository and the ability to get all repositories to find the arn needed on delete.
    new iam.Policy(this, `${this.repositoryName}Policy`, {
      roles: [role],
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'ecr:BatchDeleteImage',
            'ecr:DescribeRepositories',
            'ecr:ListImages',
            'ecr:ListTagsForResource',
          ],
          resources: [this._resource.attrArn],
        }),
      ],
    });
```

I created a stack with this code and changed the repository name and got the following error.

(And the `${repositoryName}Policy` become to `undefinedPolicy`)

```
Received response status [FAILED] from custom resource. Message returned: AccessDeniedException: User: arn:aws:sts::1234567890:assumed-role/aws-ecr-integ-stack-CustomECRAutoDeleteImagesCusto-14DD6ZRSES8VD/aws-ecr-integ-stack-CustomECRAutoDeleteImagesCusto-iDtGtDvsp9Yb is not authorized to perform: ecr:DescribeReposito
ries on resource: arn:aws:ecr:ap-northeast-1:1234567890:repository/delete-even-if-containing-images because no identity-based policy allows the ecr:DescribeRepositories action
```

> If, because of the token-ness and unpredictability of this.repositoryName, this cannot work, a good compromise may be to put a tag on the ECR repository (cdk:allow-deleting-images=true; in fact there should already be a tag on there) and use [tag-based access control](https://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr-supported-iam-actions-tagging.html) with a Condition.

In that case, is this the way to go?

Reply:
@rix0rrr 

I used `physicalName` instead of `repositoryName` and it worked!

Because `this.repositoryName` is a Token (lazy value), but `physicalName` is `props.repositoryName`.

However, `props.repositoryName` may be unspecified, in which case the `props.repositoryName` will be `undefined` and duplicate the IAM policy, which is not a good practice.

Reply:
I agree, you can't use `physicalName` in this case. We should probably go the tag route then, a naked `*` is too dangerous for my tastes.